### DISCUSSION TITLE
2D physics interpolation

### DISCUSSION THREAD

**Discussion by paulocoutinhox** [axmolengine/axmol#2091] [General]
Hi,

In latest Godot they announce "2D physics interpolation":
https://godotengine.org/releases/4.3/#highlights-2d-physics-interpolation

[physics-interpolation-2d.webm](https://github.com/user-attachments/assets/4bca12e1-fb44-467a-a393-5bc9dadc8558)

How we can do the same in Axmol?

Thanks.

**Comment by aismann** [comment]
Look at cpp-tests
Chipmunk / Box2D Testbed examples

**Comment by DelinWorks** [comment]
it's pretty simple, when you set any object position to the physics engine's position you have to keep track of the last delta time, typically in a fixed physics update it would be 60 frames a second, which is 0.0166666666666667 delta

you then store a variable `accumPhysicsDt` that every frame (game loop update) you add deltaTime to it
each time you update the physics engine you reset that variable `accumPhysicsDt` to 0.

then you map the value from `0` to the fixed delta time which is  `0.01666~` to the old object's position and the new one, keep in mind that you have to keep track of the old position which is pretty simple.

kinda like this

lastDt is `0.01666~`
`iox` and `ioy` are the old positions
`x` and `y` are the new positions

```cpp
Vec2 PhysicsWorld::getInterpolatedDynamicPosition(DynamicShape* s)
{
    float x = Math::map(accumPhysicsDt, 0.0f, lastDt, s->iox, s->x);
    float y = Math::map(accumPhysicsDt, 0.0f, lastDt, s->ioy, s->y);

    return Vec2(x, y);
}
```
you could implement it in a way where you store the future position but then you have to fiddle around with velocity and stuff that are so confusing, with this technique you have 1 frame delay because of the old position and it depends whether it's fine or not, with higher physics updates the issue becomes less of a worry.

**Reply by DelinWorks** [reply]
I see in the godot implementation they too use the 1 frame delay technique which suggests it's not really a problem
